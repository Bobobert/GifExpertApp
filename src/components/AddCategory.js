import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) => {
        e.preventDefault(); // avoids to refresh the page

        if (inputValue.trim().length > 1) {
            setCategories( categories => [inputValue, ...categories]);
            setInputValue('');
        }

    }

    return <form onSubmit={ handleSubmit }>
        <input 
            type="text"
            value={ inputValue }
            onChange={ handleInputChange }
        />
        </form>
}
//element from agroups every input element inside, so it works
// the same as a fragment in this snippet

AddCategory.propTypes = {
    setCategories : PropTypes.func.isRequired,
}