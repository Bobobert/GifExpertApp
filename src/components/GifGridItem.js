import React from 'react';

export const GifGridItem = ({id, title, url}) => {

    const handleCopy = () => {
        navigator.clipboard.writeText(`https://giphy.com/embed/${id}`).then(
        console.log(`Copied to clipboard ${title} !`));
    }

    return (
        <div className="card animate__animated animate__fadeInDown animate__fast" onClick={handleCopy}>
            <img src={ url } alt={ title }/>
            <div className='overlay'>
                <div className='text'>
                    This GIF!<br/>{title}
                </div>
            </div>

        </div>
    )
}
// extracted more info and style from this https://www.w3schools.com/css/css3_images.asp
