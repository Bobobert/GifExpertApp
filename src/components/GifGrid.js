import React from 'react'
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    //const [images, setImages] = useState([]);
    const { data:images, loading } = useFetchGifs( category );

    // useEffect(() => {
    //     getGifs();
    // }, []); // este snip sirve para bloquear la ejecucion de funciones
    // cuando son ejecutadas/ renderiadas por primera vez, al menos.
    // El problema estÃ¡ en que todo  este componente se ejecuta cuando se realiza
    // un cambio.

    // useEffect( () => {
    //     getGifs( category ).then( setImages );
    // }, [ category ]); // run just once for []
    // // or when category changes [ category ]
    
    return <>
            <h3> { category } </h3>
            { loading && <p className='animate__animated animate__shakeY'>Cargando . . .</p> }
            <div className='card-grid'>
                {images.map( img => 
                    <GifGridItem key={img.id}{...img} />
                    )}
            </div>
        </>
}
