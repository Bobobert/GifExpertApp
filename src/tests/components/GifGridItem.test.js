import { getByRole, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import React from "react";

import {GifGridItem} from "../../components/GifGridItem";

describe('Tests for the GifGridItem component', () => {

    let container = null;
    const idTest = 'A1D0';
    const titleTest = 'Test title';
    const urlTest = 'https://localhost/something.jpg';

    beforeEach(() => {
        const {container: temp} = render(<GifGridItem id={idTest} title={titleTest} url={urlTest}/>);
        container = temp; // didn't found an easier way to unpack this
    })

    afterEach(() => {
        // cleanup is already automatic.
    })

    test('simple snapshot', () => {
        expect( container ).toMatchSnapshot();
    })

    test('debe tener la imagen igual al url y alt de las propiedades', () => {
        expect( screen.getByRole('img') ).toContainHTML( titleTest );
    })

    test('debe tener la clase animate__fadeInDown', () => {

        // two ways of doing it

        expect( screen.getByTestId('GifGridItem'))
            .toHaveClass('animate__animated animate__fadeInDown');
        expect( container.firstChild )
            .toHaveClass('animate__animated animate__fadeInDown');


    }) //enzyme test things so different to the RTL, which tries to be
    // more like just an user, interacting with the DOM by Queries.

})