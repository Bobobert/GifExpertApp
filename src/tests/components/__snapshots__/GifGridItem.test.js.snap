// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the GifGridItem component simple snapshot 1`] = `
<div>
  <div
    class="card animate__animated animate__fadeInDown animate__fast"
    data-testid="GifGridItem"
  >
    <img
      alt="Test title"
      src="https://localhost/something.jpg"
    />
    <div
      class="overlay"
    >
      <div
        class="text"
      >
        This GIF!
        <br />
        Test title
      </div>
    </div>
  </div>
</div>
`;
